lib/ansible/runner.rb

```diff
@@ -254,6 +254,12 @@ def runner_params(base_dir, ansible_runner_method, playbook_or_role_args, verbos
         runner_args[:role_skip_facts] = nil if runner_args.delete(:role_skip_facts)
         runner_args[:ident] = "result"
 
+        playbook = runner_args.delete(:playbook)
+        if playbook
+          runner_args[:playbook]    = File.basename(playbook)
+          runner_args[:project_dir] = File.dirname(playbook)
```

Comment:
BTW, recently we determined that we may have to support playbooks in subdirectories, so this code may end up being invalid (along with a couple of other places)

Reply:
@Fryguy for what it is worth, I have taken said BZ, and don't think it will be a problem.

I see this as when it gets to the `lib/ansible/runner.rb` point, we would have already passed it the subdirectory for the path, and not the top level directory of the clone.  Will get back to you on that when I open up a PR, but I don't think we will need to worry about that now.

Reply:
I think @Fryguy is right here. If we pass a playbook nested deeply in a repo then based on https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse_roles.html#role-search-path ansible will look for the roles directory as a peer of the playbook, but I'm not sure this would have worked previously ...

Reply:
Oh, I just realized that you weren't talking about roles...

This should work fine for running a playbook without roles, but if we had installed the roles into to top-level `roles` directory in the repo then the playbook run probably wouldn't find them.

Reply:
@carbonin I am pretty sure he was referring to this:

https://gitter.im/ManageIQ/manageiq-providers-embedded_ansible?at=5d4048e4a9200c242824a7d4
https://bugzilla.redhat.com/show_bug.cgi?id=1734446

Where this repo:

https://gitlab.com/redhat-cop/mbu-lab/ansible

Would not work.  I think the way we solve this is we pass `/path/to/{{repo_dir}}/Playbooks` for that for our repo directory for runner (instead of the top level of the repo), and what you have here should work fine as a result.

* * *

I will be looking into what `awx` does to determine this for sure, but I don't think it is something we need to solve in this PR, but something I will keep in mind when I take that on. By the sounds of things, since this is used by the `mbu-lab`, this playbook probably worked in some capacity in the past, but we (I) should probably double check that is the case as well.

Reply:
Yes, I was referring to what @NickLaMuro said, but I'm also extrapolating to the known code paths where we make an assumption of the playbook being in the root dir.  Namely this line of code (via .basename and .dirname is the assumption) and also https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/embedded_ansible/automation_manager/configuration_script.rb#L37

Reply:
Ah okay. So as far as this patch goes, that should be fine.

Now if any of those repos looked like this:
```
/roles
  requirements.yml
/things
  /stuff
    playbook.yml (that uses something in `requirements.yml`)
```

Then we'd probably have an issue, but that's separate from what you're trying to solve.

Reply:
@carbonin Okay, now that I have gotten to look at what `ansible/awx` is doing, I think what I said above was wrong (sorry 😞 ) and we should change how we are doing this.

So the code that "finds playbooks" in a project is here:

https://github.com/ansible/awx/blob/128fa8947ac620add275a15cb07577178745a849/awx/main/models/projects.py#L191-L202

And as your ansible doc link from above stated:

> Role Search Path
> ----------------
> 
> Ansible will search for roles in the following way:
> 
> * A `roles/` directory, relative to the playbook file.
> * By default, in `/etc/ansible/roles`

Given that, it seems like tower is following that rule to the letter and roles are playbook relative, not "where my playbooks happen to live" relative (sorry for missing that bit earlier).  Basing this off of the default vars passed to runner in `tasks.py` in `awx`:

https://github.com/ansible/awx/blob/1242ee2b/awx/main/tasks.py#L1766-L1780

So for the given directory structure above:

```
roles/
  requirements.yml
things/
  stuff/
    playbook.yml (that uses something in `requirements.yml`)
```

I think the `roles/` directory in the above will always end up being skipped over in all default ansible scenarios when running `things/stuff/playbook.yml`, unless the `role_path` option is specifically passed.  I think something more reasonable that we would see in the wild is:

```
azure_servers/
  roles/
    azure_stuffs/
    requirements.yml
  azure_playbook.yml
aws_servers/
  roles/
    s3_stuffs/
    ec2_config/
  aws_playbook.yml
  s3_playbook.yml
localhost/
  requirements.yml
  localhost_playbook.yml
```

So the roles are contextual, and we should only be installing the requirements for the roles that would be associated with the playbook being run.  Unsure if the `localhost/requirements.yml` is best practice, but that is the case where if a `localhost/roles/requirements.yml` were used, that file would be favored instead.

For your code change above, I thing it instead makes more sense for the `Ansible::Content.new(my_path).fetch_galaxy_roles` that is in `ConfiguationScriptSource` to instead live here, and with `runner_args[:project_dir]` as the directory passed into `Ansible::Content.new` as well instead of the checkout location's root path.

There possibly an edge case, where if the user does override the `role_path` via `extra_vars` or the `env`, this would probably need to be updated, but I think we can address that later.

* * *

Then  to fix the BZ referenced in Jason's comment:

https://bugzilla.redhat.com/show_bug.cgi?id=1734446

the change that I would end up putting together will just be to update the `#playbooks_in_git_repository` of `ConfigurationScriptSource` to be "smarter" about finding playbooks in nested directories, and that should be it, so this PR and those changes should remain independent of each other (though, we need to agree that what I just babbled about is correct...).

Reply:
:+1: I think this is the right way to go.